pkg/kv/kvserver/rangefeed/buffered_sender.go

```diff
@@ -137,13 +141,63 @@ func (bs *BufferedSender) sendBuffered(
 	if bs.queueMu.stopped {
 		return errors.New("stream sender is stopped")
 	}
-	if bs.queueMu.overflowed {
-		return newRetryErrBufferCapacityExceeded()
-	}
-	if bs.queueMu.capacity > 0 && bs.queueMu.buffer.len() >= bs.queueMu.capacity {
-		bs.queueMu.overflowed = true
-		return newRetryErrBufferCapacityExceeded()
+
+	// Per-stream capacity limits. If the stream is already overflowed we drop the
+	// request. If the stream has hit its limit, we return an error to the
+	// registration. This error should be the next event that is sent to
+	// stream.
+	status := bs.queueMu.byStream[ev.StreamID]
+	switch status.state {
+	case streamActive:
+		if bs.queueMu.perStreamCapacity > 0 && status.queueItems == bs.queueMu.perStreamCapacity {
+			if ev.Error != nil {
+				// If _this_ event is an error, no use sending another error. This stream
+				// is going down. Admit this error and mark the stream as overflowed.
+				status.state = streamOverflowed
+			} else {
+				// This stream is at capacity, return an error to the registration that it
+				// should send back to us after cleaning up.
+				//
+				// TODO(review): We could alternatively put the error event onto the
```

Comment:
I'm OOTL on this code these days so feel free to just ignore me here -- I completely trust y'all know what you're doing here better than me. I don't quite follow the two-step overflow transition? why we do we need to wait for the sender to come back and give us an error to enqueue vs do it here ourselves? I don't quite follow what stronger assertion we get by doing this.

Reply:
Thanks for taking a look - I appreciate extra eyes on this since I’m often unsure about what is is right here. 

We currently have three ways a stream can disconnect: when a registration hits an error, when the buffered sender shuts down all streams via the stream manager, and when there’s a close stream request. I already feel like that’s too many ways. If we let the buffered sender disconnect a single stream directly, we’d need to call both Disconnect and onError clean up ourselves. 

So I kind of like the current approach where we only call Disconnect and let the error pass through to the queue, similar to how CloseStream works although I agree with you that the overflowing -> overflowed transition is annoying.

Reply:
I'm not really sure it is a stronger assertion. But, the "unbuffered registration" sits between the processor (which sends events for a range) and this buffered sender.  One nice property of waiting for the error is that then when we take the error out of the buffer, we know that the registration has been marked as disconnected and run any cleanup. In general, having all errors go first to the registration has really helped me think about the lifecycle of this code so I think that is why I tried to preserve this here.